vsr: add reconfiguration request

matklad opened the pull request:
This adds the basic boilerplate to handle reconfiguration:

   - `.reconfigure` operation
   - Corresponding `ReconfigurationRequest` which forms the body of the request.
   - Basic validation logic (the bit where primary makes a judgement call whether reconfiguration is a good idea based on status of the cluster is yet TODO)
   - Skeleton of request handling by the primary (during prepare and commit).

The `ReconfigurationRequest` is somewhat future proof --- it includes facilities for changing the number of standbys/actives, although we don't support actual changes here yet (only permutations are allowed). I _think_ we could re-use the same request shape for both:

* logical reconfiguration (this PR; advance epoch, permute members, but keep the member set the same).
* physical reconfiguration (keep the epoch and active replicas, but add or remove a standby).

matklad committed:
vsr: add reconfiguration request

This adds the basic boilerplate to handle reconfiguration:

   - `.reconfigure` operation
   - Corresponding `ReconfigurationRequest` which forms the body of the
     request.
   - Basic validation logic (the bit where primary makes a judgement
     call whether reconfiguration is a good idea based on status of the
     cluster is yet TODO)
   - Skeleton of request handling by the primary (during prepare and
     commit).

The `ReconfigurationRequest` is somewhat future proof --- it includes
facilities for changing the number of standbys/actives, although we
don't support actual changes here yet (only permutations are allowed).
I _think_ we could re-use the same request shape for both:

* logical reconfiguration (this PR; advance epoch, permute members, but
  keep the member set the same).
* physical reconfiguration (keep the epoch and active replicas, but add
  or remove a standby).

matklad commented on src/vsr.zig:
```diff
@@ -142,6 +142,8 @@ pub const Operation = enum(u8) {
     /// The value 2 is reserved to register a client session with the cluster.
     register = 2,
+    /// The value 3 is reserved for reconfiguration request.
+    reconfigure = 3,
```
Didn't find a better name for the request here, the problem with reconfiguration is that configuration is too generic. On the plus side, this maintains the invariant that all reserved operations start with r!

sentientwaffle replied:
Brainstorming: shape/reshape; roster; roll.

(Roll is a pun — "roll" as in a list of names/items, or "roll" as in circular, i.e. a ring — which is not a reason to use it, but is funny anyway :grin:)

matklad replied:
From today's chat with Joran --- an important point here is that this is our public visible interface, rather than just our private API. In the docs, we do want to call this "reconfiguration" as that's the appropriate term here.

sentientwaffle replied:
Good point; I hadn't considered that.

matklad commented on src/vsr.zig:
```diff
@@ -1189,13 +1417,20 @@ test "quorums" {
+/// First replica_count elements are active replicas,
+/// then standby_count standbys, the rest are zeros.
+/// Order determines ring topology for replication.
+pub const Members = [constants.nodes_max]u128;
```
I am wondering if I should wrap this in a struct just to have something to put methods to, instead of doing c-style `members_this`, `members_that`. Decided to stick to C-style, as that's simpler

matklad commented on src/vsr/replica.zig:
```diff
@@ -4637,6 +4672,23 @@ pub fn ReplicaType(
+                vsr.ReconfigurationRequest,
+                request.body()[0..@sizeOf(vsr.ReconfigurationRequest)],
+            );
+            reconfiguration_request.*.result = reconfiguration_request.validate(.{
```
Here, we curiously take advantage of our ability to do some computation during prepare, rather than commit. 

For reconfiguration, primary needs to make a decision whether its good or not during prepare, not during commit; preparing reconfigration takes effect immediately on the primary, as it must stop processing any further requests.

sentientwaffle commented on src/stdx.zig:
```diff
@@ -216,6 +216,19 @@ pub inline fn hash_inline(value: anytype) u64 {
+/// Intended exclusively for table-driven prototype-based tests. Write
+/// updates explicitly in production code.
+pub fn update(base: anytype, diff: anytype) @TypeOf(base) {
+    assert(builtin.is_test);
```
Maybe add `assert(@TypeOf(base) == .Struct);` -- that way passing a pointer for `base` will fail rather than mutating.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+    }
+};
+
+pub const ReconfigurationResult = enum(u32) {
```
I know we don't do this now for other protocol enums, but what do you think of setting the values explicitly?
e.g. `enum { ok = 0, replica_count_zero = 1, ... }`.

Motivation: Make it extra obvious that renumbering is a breaking protocol change, i.e. new operations should be appended.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+    /// The result of this request. Set to zero by the client and filled-in by the primary when it
+    /// accepts a reconfiguration request.
+    result: ReconfigurationResult,
+
```
```suggestion
comptime {
    assert(@bitSizeOf(ReconfigurationRequest) == @sizeOf(ReconfigurationRequest) * 8);
}
```
Verify there is no implicit padding.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+        assert(valid_members(current.members));
+        assert(member_count(current.members) == current.replica_count + current.standby_count);
+
+        if (request.replica_count == 0) return .replica_count_max_exceeded;
```
```suggestion
        if (request.replica_count == 0) return .replica_count_zero;
```

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+        if (request.replica_count != current.replica_count) return .different_replica_count;
+        if (request.standby_count != current.standby_count) return .different_standby_count;
+
+        // Note: neither set can have duplicates at this point.
```
Do you mean that it is impossible for there to be duplicates at this point?
Or that if there are duplicates, then it is an error?
(I'm think it is the former, but I think this could be made more clear.)

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+        .epoch_in_the_future,
+    );
+
+    t.epoch = 0xFFFF_FFFF;
```
`std.math.maxInt(u32)` might make the test's intent slightly more clear, but either is fine.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+    /// At the moment, we require this to be equal to the old count.
+    standby_count: u8,
+    reserved: [10]u8 = [_]u8{0} ** 10,
+    /// The result of this request. Set to zero by the client and filled-in by the primary when it
+    /// accepts a reconfiguration request.
+    result: ReconfigurationResult,
```
I wonder if `0` (set by client) should correspond to a special `ReconfigurationResult` (i.e. not `ok`)?
e.g. `request`.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+            } else return .different_member_set;
+        }
+
+        if (std.meta.eql(request.members, current.members.*)) return .no_op;
```
Maybe use `std.mem.eql` instead, to avoid copying the arrays onto the stack.
```suggestion
        if (std.mem.eql(u128, &request.members, current.members)) return .no_op;
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -3344,6 +3346,36 @@ pub fn ReplicaType(
+            const reconfiguration_request = std.mem.bytesAsValue(
+                vsr.ReconfigurationRequest,
+                prepare.body()[0..@sizeOf(vsr.ReconfigurationRequest)],
+            );
```
```suggestion
            );
            assert(reconfiguration_request.result != .reserved);
```

sentientwaffle replied:
Did this assert get missed? I see it in `primary_prepare_reconfiguration` but I think it applies to `commit_reconfiguration` too.

matklad replied:
indeed, added!

sentientwaffle commented on src/vsr.zig:
```diff
@@ -811,6 +817,271 @@ pub const BlockRequest = extern struct {
+        .configuration_is_no_op,
+    );
+
+    assert(t.tested.count() < ResultSet.initFull().count());
+    t.tested.insert(.reserved);
+    assert(t.tested.count() == ResultSet.initFull().count());
```
Nice coverage check!

matklad commented:
Status: waiting for @jorangreef to take a look as well

matklad committed:
vsr: refactor away epoch_overflow error

matklad committed:
vsr: save vertical space a bit

I tried an `inline for` over `test_cases` table here, but that sadly
segfaults the compiler at the moment.

matklad commented:
Walked through the PR with jorangreef, should be good to go now. We might still want to work on naming for various error variants a bit, but that's related to `ok_created_already` discussion I also have in flight ....

sentientwaffle reviewed (approved):
LGTM, though maybe in a follow-up add a comptime assert to `ReconfigurationResult` to ensure that there are no skipped variant values in the enum:

```
    comptime {
        for (std.enums.values(ReconfigurationResult)) |result, index| {
            assert(@enumToInt(result) == index);
        }
    }
```
